crates/rust/src/lib.rs

```diff
@@ -1272,6 +1281,40 @@ fn bitcast(casts: &[Bitcast], operands: &[String], results: &mut Vec<String>) {
             Bitcast::I64ToF64 => format!("f64::from_bits({} as u64)", operand),
             Bitcast::F32ToI64 => format!("i64::from(({}).to_bits())", operand),
             Bitcast::I64ToF32 => format!("f32::from_bits({} as u32)", operand),
+
+            // Convert an `i64` into a `MaybeUninit<u64>`.
+            Bitcast::I64ToP64 => format!("::core::mem::MaybeUninit::new({} as u64)", operand),
+            // Convert a `MaybeUninit<u64>` holding an `i64` value back into
+            // the `i64` value.
+            Bitcast::P64ToI64 => format!("{}.assume_init() as i64", operand),
+
+            // Convert a pointer value into a `MaybeUninit<u64>`.
+            Bitcast::PToP64 => {
+                format!(
+                    "{{
+                        let mut t = ::core::mem::MaybeUnunit::<u64>::uninit();
+                        t.as_mut_ptr().cast::<*mut core::ptr::c_void>().write({});
+                        t
+                    }}",
+                    operand
+                )
+            }
+            // Convert a `MaybeUninit<u64>` holding a pointer value back into
+            // the pointer value.
+            Bitcast::P64ToP => {
+                format!(
+                    "{}.as_mut_ptr().cast::<*mut core::ptr::c_void>().read()",
+                    operand
+                )
+            }
+            // Convert an `i32` into a pointer.
+            Bitcast::I32ToP => {
+                format!("{} as *mut ::core::ffi::c_void", operand)
+            }
```

Comment:
Could this be removed entirely? It seems like we should avoid generating this ever because this implies that provenance can be manifest out of nowhere right?

Reply:
(and similarly below, could `PToI32` be removed entirely?

Reply:
When we have a variant of an `i32` and a pointer, we unify the two types to a pointer type, so that when the variant does hold a pointer, it preserves provenance. When the variant holds an i32, there is no provenance anyway, so it's ok that we're doing these conversions without preserving provenance.

Reply:
Aha I see, makes sense!

Could you add some docs with respect to provenance on the `Bindgen` enum variants? If I understand right it's intended that `I32ToP` adds arbitrary provenance and `PToI32` removes that arbitrary provenance, right?

Reply:
I've now added some docs.

The way it's discussed in the Rust provenance docs, they say that what `I32ToP` does is creates a pointer with no provenance. Once the API is stablized in Rust, the functions that will replace these `as` casts have names like [`without_provenance`](https://github.com/rust-lang/rust/pull/117658).